src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import category_theory.arrow
+import category_theory.limits.preserves.shapes.terminal
+
+/-!
+# Lifting properties
+
+This file defines the lifting property of two arrows in a category and shows basic properties of
+this notion.
+We also construct the subcategory consisting of those morphisms which have the right lifting
+property with respect to arrows in a given diagram.
+
+## Main results
+- `has_lifting_property`: the definition of the lifting property
+- `iso_has_right_lifting_property`: any isomorphism satisfies the right lifting property (rlp)
+- `id_has_right_lifting_property`: any identity has the rlp
+- `right_lifting_property_initial_iff`: spells out the rlp with respect to a map whose source is an
+initial object
+- `right_lifting_subcat`: given a set of arrows `F : D → arrow C`, we construct the subcategory
+of those morphisms `p` in `C` that satisfy the rlp w.r.t. `F i`, for any element `i` of `D`.
+
+## Tags
+lifting property
+-/
+
+open category_theory.limits
+
+namespace category_theory
+
+universes v u v₁
+variables {C : Type u} [category.{v} C]
+variables {D : Type v₁}
+
+variables {X Y Z : C}
+
+/-- The lifting property of a morphism `i` with respect to a morphism `p`.
+This can be interpreted as the right lifting property of `i` with respect to `p`,
+or the left lifting property of `p` with respect to `i`. -/
+class has_lifting_property (i p : arrow C) : Prop :=
+(sq_has_lift : ∀ (sq : i ⟶ p), arrow.has_lift sq)
+
+@[priority 100] -- see Note [lower instance priority]
+instance has_lifting_property' {i p : arrow C} [has_lifting_property i p] (sq : i ⟶ p) :
+  arrow.has_lift sq :=
+has_lifting_property.sq_has_lift sq
+
+/-- Any isomorphism has the right lifting property with respect to any map.
+A    → X
+↓i    ↓p≅
+B    → Y
+-/
+lemma iso_has_right_lifting_property (i : arrow C) (p : X ≅ Y) :
+  has_lifting_property i (arrow.mk p.hom) :=
+⟨λ sq, ⟨⟨{ lift := sq.right ≫ p.inv, }⟩⟩⟩ -- the lift is obtained by p⁻¹ ∘ (B → Y)
+
+/-- Any identity has the right lifting property with respect to any map. -/
+lemma id_has_right_lifting_property (i : arrow C) : has_lifting_property i (arrow.mk (𝟙 X)) :=
+  iso_has_right_lifting_property i (iso.refl _)
```

Comment:
I guess the style can be confusing initially, but the idea is as follows:

  * the *type* is indented (so whatever is between `:` and `:=`) if it starts on a new line (in this case it doesn't which is fine)
  * the *term* (so the stuff after `:=`) is flushleft in column 0

```suggestion
iso_has_right_lifting_property i (iso.refl _)
```

Reply:
OK.